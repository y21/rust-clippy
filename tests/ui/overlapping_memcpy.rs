#![allow(unused)]
#![warn(clippy::overlapping_memcpy)]

fn main() {
    let mut s = [0; 100];
    let ptr = s.as_mut_ptr();
    unsafe {
        std::ptr::copy_nonoverlapping(ptr, ptr, 10);
        std::ptr::copy_nonoverlapping(ptr, ptr.add(9), 10);
        std::ptr::copy_nonoverlapping(ptr, ptr.add(10), 10);
        std::ptr::copy_nonoverlapping(ptr.add(10), ptr, 10);
        std::ptr::copy_nonoverlapping(ptr.add(10), ptr, 11);
        std::ptr::copy_nonoverlapping(ptr.add(10), ptr, 9);
        std::ptr::copy_nonoverlapping(ptr.add(10), ptr.add(10), 10);
    }
}
