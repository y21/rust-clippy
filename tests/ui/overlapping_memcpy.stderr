error: calling `ptr::copy_nonoverlapping` with overlapping regions
  --> $DIR/overlapping_memcpy.rs:8:9
   |
LL |         std::ptr::copy_nonoverlapping(ptr, ptr, 10);
   |         -----------------------------^^^^^^^^^^^^^^
   |         |
   |         help: instead use: `std::ptr::copy`
   |
note: source pointer covers range `0..10`
  --> $DIR/overlapping_memcpy.rs:8:39
   |
LL |         std::ptr::copy_nonoverlapping(ptr, ptr, 10);
   |                                       ^^^
note: destination pointer covers range `0..10`
  --> $DIR/overlapping_memcpy.rs:8:44
   |
LL |         std::ptr::copy_nonoverlapping(ptr, ptr, 10);
   |                                            ^^^
   = note: it is Undefined Behavior to call `ptr::copy_nonoverlapping` with overlapping regions
   = note: `-D clippy::overlapping-memcpy` implied by `-D warnings`

error: calling `ptr::copy_nonoverlapping` with overlapping regions
  --> $DIR/overlapping_memcpy.rs:14:9
   |
LL |         std::ptr::copy_nonoverlapping(ptr.add(10), ptr.add(10), 10);
   |         -----------------------------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |         |
   |         help: instead use: `std::ptr::copy`
   |
note: source pointer covers range `10..20`
  --> $DIR/overlapping_memcpy.rs:14:39
   |
LL |         std::ptr::copy_nonoverlapping(ptr.add(10), ptr.add(10), 10);
   |                                       ^^^^^^^^^^^
note: destination pointer covers range `10..20`
  --> $DIR/overlapping_memcpy.rs:14:52
   |
LL |         std::ptr::copy_nonoverlapping(ptr.add(10), ptr.add(10), 10);
   |                                                    ^^^^^^^^^^^
   = note: it is Undefined Behavior to call `ptr::copy_nonoverlapping` with overlapping regions

error: aborting due to 2 previous errors

